### PR TITLE
Made all exceptions inherit from a base exception

### PR DESCRIPTION
**Pull request by poisa**
Implements changes proposed in #115.

> Currently you have to catch every single exception this package throws. If you make a generic exception, say `Google2FAException` and have all other exceptions extend this one it is much easier for consumers of this package to catch anything thrown by it. Sometimes you don't care exactly what exception was thrown but you'd like to identify the general area (or package) of the problem.

### Changes in tests:

Unit tests used the `@expectedException` to test for exceptions. There is nothing wrong with this approach except that we now want to test two things: 

1) That the exception is a subclass of `Google2FAException`
2) That the exception is an instance of each particular exception.

In order to achieve both, I wrapped the test calls in a try/catch block which catches with the generalized exception, and then checks for the specific exception.

### CODE REVIEW

**Comment by GrahamCampbell** [comment]
It is more usual to have exceptions implement a `Google2FAExceptionInterface`, so we are not forced to expend only `Exception`, and could later choose to extend other classes such as `InvalidArgumentException`.

**Comment by antonioribeiro** [comment]
Nice, @poisa, thanks for taking time to make this happen! 
Can we go to the direction @GrahamCampbell is pointing here?

**Comment by antonioribeiro** [comment]
@poisa, doing some tests on it, all those catches are preventing tests from failing, which is bad, really bad. 

Tests are supposed to fail if we insert a bug in the code, but if you catch and handle it, they won't fail as they are supposed, right?

``` php
public function testShortSecretKey()
{
    $this->google2fa->setEnforceGoogleAuthenticatorCompatibility(false);

     try {
        $this->google2fa->verifyKey(
            Constants::SHORT_SECRET,
            'THIS IS A BUG', // <------------- BUG
            null,
            26213400
        );
    } catch (Google2FAException $exception) {
        $this->assertInstanceOf('PragmaRX\Google2FA\Exceptions\SecretKeyTooShortException', $exception);
    }
}
```

**Comment by poisa** [comment]
Implemented changes suggested by @GrahamCampbell.

@antonioribeiro you are right but that is not a byproduct of my commit. That test in particular needs reworking as the test case you suggested fails in the master branch too.

**Comment by antonioribeiro** [comment]
> @antonioribeiro you are right but that is not a byproduct of my commit. That test in particular needs reworking as the test case you suggested fails in the master branch too.

They are passing nicely here

image

An on Travis: https://travis-ci.org/antonioribeiro/google2fa/builds

**Comment by poisa** [comment]
Sorry, what meant to say was that the test you suggested, that is, adding
the string “THIS IS A BUG” as a parameter PASSES in master too. So the
problem has nothing to do with my commit but with that test or with the
actual code which are not producing an exception for that particular case.

Do this:
- Check out master.
- Introduce the same bug you suggested.
- Run the tests.
- The tests will pass.

This proves that the issue was there before any of my commits.

**Comment by antonioribeiro** [comment]
Yes the tests are passing but they are not supposed to pass since I introduced a bug. But as you are catching those errors, the catch is doing the job of handling an exception it was not supposed to handle. 

Tests are meant to fail when we introduce bugs in tests or in our app/package, but if you handle them, we won't see bugs in the app/package.

So, for this bug to show up to us:

``` php
public function testShortSecretKey()
{
    $this->google2fa->setEnforceGoogleAuthenticatorCompatibility(false);

     try {
        $this->google2fa->verifyKey(
            Constants::SHORT_SECRET,
            'THIS IS A BUG', // <------------- BUG
            null,
            26213400
        );
    } catch (Google2FAException $exception) {
        $this->assertInstanceOf('PragmaRX\Google2FA\Exceptions\SecretKeyTooShortException', $exception);
    }
}
```

We MUST remove the try/catch:

``` php
public function testShortSecretKey()
{
    $this->google2fa->setEnforceGoogleAuthenticatorCompatibility(false);

    $this->google2fa->verifyKey(
        Constants::SHORT_SECRET,
        'THIS IS A BUG', // <------------- BUG
        null,
        26213400
    );
}
```

**Comment by antonioribeiro** [comment]
I don't really know why you introduced all those try/catch in your code, because I just pulled your PR and replaced all tests by the ones we have on master and it is passing like a sharm!

**Comment by antonioribeiro** [comment]
@GrahamCampbell, is this concept you were talking about? https://www.brandonsavage.net/using-interfaces-for-exceptions/, I mean, implementing many different contracts to a base exception?

**Comment by poisa** [comment]
>Yes the tests are passing but they are not supposed to pass since I introduced a bug. But as you are catching those errors, the catch is doing the job of handling an exception it was not supposed to handle.

I can't be catching any errors since I was talking about the `master` branch where I wrote no code. If you introduce that bug, the tests in master pass when they should fail. This is an _pre-existing_ issue that is out of the scope of my code.

>I don't really know why you introduced all those try/catch in your code, 

This is explained in the PR description but I will explain it differently in the following paragraphs.

>I just pulled your PR and replaced all tests by the ones we have on master and it is passing like a charm!

They pass, but they are not testing what they should. The tests in master make sure the code throws the particular exceptions `InvalidCharactersException`, `IncompatibleWithGoogleAuthenticatorException`, and `SecretKeyTooShortException`. The tests in my PR test that not only those exceptions are thrown, but also that they implement the interface in cuestion. If you check my PR out and remove `implements Google2FAExceptionInterface` from one of the exceptions classes you will see that the tests fail.

I think you are misunderstanding how PHPUnit works internally. Let me do a bird's eye view of the process so that we are both talking about the same thing.

Your code throws an exception so you write a test to trigger that exception and have PHPUnit check that it was thrown. One way of doing this is by using annotations like you did. So you can add the `@expectedException` or you can add the line `$this->expectException([...])` in your test. This tells PHPUnit to mark the test as passing if this particular exception is thrown. The way it does that is by wrapping each test with a try/catch block and comparing the thrown exception -if any- with the one you expected. 

But there is one thing the above method can't do: test more than one thing in the exception. This is exactly what we want to do as we want to test that the particular exception is thrown and that it implements the interface. One way of doing that is what I did, catch the *interface* inside the test and then make sure the caught exception is of the particular type we want to see by way of an assertion. If this test passes then it means that the code has indeed thrown the exception (and that it implements the interface). If the exception is not thrown then PHPUnit will mark the test as `risky` and you will see a yellow warning message in the output. If you wanted the test to actually *fail*, we could do this:

```php
    public function testShortSecretKey()
    {
        $this->google2fa->setEnforceGoogleAuthenticatorCompatibility(false);

        $exceptionWasThrown = false;

        try {
            $this->google2fa->verifyKey(
                Constants::SHORT_SECRET,
                '558854',
                null,
                26213400
            );
        } catch (Google2FAExceptionInterface $exception) {
            $this->assertInstanceOf('PragmaRX\Google2FA\Exceptions\SecretKeyTooShortException', $exception);
            $exceptionWasThrown = true;
        }

        $this->assertTrue($exceptionWasThrown, 'Expected exception but none was thrown.');
    }
```

If the exception is not thrown or does not implement the interface, then the test will be marked as `failed` instead of `risky`.

Another way to test this using annotations would be to copy the test over and use a different annotation on each, like this:

```php
    /**
     * @expectedException \PragmaRX\Google2FA\Exceptions\SecretKeyTooShortException
     */
    public function testShortSecretKeyThrowsSecretKeyTooShortException()
    {
        $this->google2fa->setEnforceGoogleAuthenticatorCompatibility(false);

        $this->google2fa->verifyKey(
            Constants::SHORT_SECRET,
            '558854',
            null,
            26213400
        );
    }

    /**
     * @expectedException \PragmaRX\Google2FA\Exceptions\Google2FAExceptionInterface
     */
    public function testShortSecretKeyThrowsGoogle2FAExceptionInterface()
    {
        $this->google2fa->setEnforceGoogleAuthenticatorCompatibility(false);

        $this->google2fa->verifyKey(
            Constants::SHORT_SECRET,
            '558854',
            null,
            26213400
        );
    }
```

So to recap,

1) The existing code that is in the master branch passes all the tests even if you include the bug. This is unrelated to my code and you would have to fix this yourself.
2) My PR marks the tests as risky if the exceptions are not thrown. **If you want I can update it to one of the above mentioned solutions so that tests actually fail.** 
3) Using the existing tests with my code would not work as they are not testing the new interface but only the already existing exceptions.
4) There is no harm in using try/catch in tests as that is exactly what PHPUnit does internally.

**Comment by antonioribeiro** [comment]
Nice, thanks for the explanation.

I refactored a bit your code to add (back) the base Exception class, but also add interfaces for all of our exceptions and a Contracts folder to hold them in a proper namespace. This way we should be covered for those who like handle interfaces better than Exception classes and vice-versa, also giving us the option to move forward. And if we introduce a new kind of Exception, we can still use interfaces to correctly type it.

I also refactored tests to make them a little bit more readable, by splitting those catches into several new test methods.

@poisa && @GrahamCampbell, could you please check it and tell me what you think about it now?

Tests are passing here and in Travis: https://travis-ci.org/antonioribeiro/google2fa. But I removed PHP 5.4 and 5.5 from the test stack.

I don't think this will introduce breaking changes, so I'm thinking about tagging a 5.1.0 version, but I also would like your input about this.

Thank you very much.
